yt_dlp/networking/_curlcffi.py

```diff
@@ -0,0 +1,216 @@
+
+curl_cffi_version = tuple(int_or_none(x, default=0) for x in curl_cffi.__version__.split('.'))
+
+if curl_cffi_version < (0, 5, 10):
+    raise ImportError('Only curl_cffi>=0.5.10 is supported')
+
+import curl_cffi.requests
+from curl_cffi.const import CurlECode, CurlOpt
+
+
+class CurlCFFIResponseReader(io.IOBase):
+    def __init__(self, response: curl_cffi.requests.Response):
+        self._response = response
+        self._buffer = b''
+        self._eof = False
+        self.bytes_read = 0
+
+    def readable(self):
+        return True
+
+    def read(self, size=None):
+        exception_raised = True
+        try:
+            while not self._eof and (size is None or len(self._buffer) < size):
+                chunk = next(self._response.iter_content(), None)
+                if chunk is None:
+                    self._eof = True
+                    break
+                self._buffer += chunk
+                self.bytes_read += len(self._buffer)
+
+            if size is None:
+                data = self._buffer
+                self._buffer = b''
+            else:
+                data = self._buffer[:size]
+                self._buffer = self._buffer[size:]
+
+            # "free" the curl instance if the response is fully read.
+            # curl_cffi doesn't do this automatically and only allows one open response per thread
+            if self._eof and len(self._buffer) == 0:
+                self.close()
+            exception_raised = False
+            return data
+        finally:
+            if exception_raised and not self.closed:
+                self.close()
+
+    def close(self):
+        self._response.close()
+        super().close()
+
+
+class CurlCFFIResponseAdapter(Response):
+    fp: CurlCFFIResponseReader
+
+    def __init__(self, response: curl_cffi.requests.Response):
+        super().__init__(
+            fp=CurlCFFIResponseReader(response),
+            headers=response.headers,
+            url=response.url,
+            status=response.status_code)
+
+    def read(self, amt=None):
+        try:
+            return self.fp.read(amt)
+        except curl_cffi.requests.errors.RequestsError as e:
+            if e.code == CurlECode.PARTIAL_FILE:
+                content_length = int_or_none(e.response.headers.get('Content-Length'))
+                raise IncompleteRead(
+                    partial=self.fp.bytes_read,
+                    expected=content_length - self.fp.bytes_read if content_length is not None else None,
+                    cause=e) from e
+            raise TransportError(cause=e) from e
+
+
+@register_rh
+class CurlCFFIRH(ImpersonateRequestHandler, InstanceStoreMixin):
+    RH_NAME = 'curl_cffi'
+    _SUPPORTED_URL_SCHEMES = ('http', 'https')
+    _SUPPORTED_FEATURES = (Features.NO_PROXY, Features.ALL_PROXY)
+    _SUPPORTED_PROXY_SCHEMES = ('http', 'https', 'socks4', 'socks4a', 'socks5', 'socks5h')
+    _SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP = {
```

Comment:
Re: https://github.com/yt-dlp/yt-dlp/pull/7595/files#r1412776681

We are hard-coding the list here anyway. We just need to refactor the code a bit so the keys are available even when curl_cffi isn't imported

Reply:
I'm not a fan of that. Also it's not guaranteed we know this ahead of time, so it doesn't generalize. Just in this very particular case it is. 

It is dependent on whatever the user has installed and what that supports, and so to get a complete list users are intended to consult the creator.